cpp/src/arrow/dataset/dataset_writer_test.cc

```diff
@@ -234,6 +234,18 @@ TEST_F(DatasetWriterTestFixture, BasicFileDirectoryPrefix) {
   AssertFilesCreated({"testdir/a/1_chunk-0.arrow"});
 }
 
+TEST_F(DatasetWriterTestFixture, DirectoryCreateFails) {
+  // This should fail to be created
+  write_options_.base_dir = "///doesnotexist";
+  EXPECT_OK_AND_ASSIGN(auto dataset_writer,
+                       DatasetWriter::Make(write_options_, scheduler_.get()));
+  Future<> queue_fut = dataset_writer->WriteRecordBatch(MakeBatch(100), "a", "1_");
+  AssertFinished(queue_fut);
+  ASSERT_OK(dataset_writer->Finish());
+  scheduler_->End();
+  ASSERT_FINISHES_AND_RAISES(Invalid, scheduler_->OnFinished());
```

Comment:
Hmm... so the only way to know that writing failed is to end the scheduler? Why isn't it popped up earlier?